### ISSUE TITLE
multi gpus problem

### ISSUE THREAD

**Issue by marvis** [AlexeyAB/darknet#1098]
Hi AlexeyAB,

When training with multi-gpus, the learning rate and batch size will increase ngpus times, I wonder why the num_batches doesn't have to decrease to 1/ngpus.

For examples, when train yolov3 on coco
one gnu: ./darknet detector train cfg/coco.data cfg/yolov3.cfg darknet53.conv.74

four gpus: ./darknet detector train cfg/coco.data cfg/yolov3.cfg darknet53.conv.74 -gpus 0,1,2,3

It is unfair that 4gpus will pass more data then 1gpu.

Can you answer the question? Thanks!

**Comment by AlexeyAB** [comment]
@marvis Hi,

> When training with multi-gpus, the learning rate and batch size will increase ngpus times

Yes: https://github.com/pjreddie/darknet/blob/f6d861736038da22c9eb0739dca84003c5a5e275/examples/detector.c#L27
https://github.com/pjreddie/darknet/blob/f6d861736038da22c9eb0739dca84003c5a5e275/examples/detector.c#L32

>  I wonder why the num_batches doesn't have to decrease to 1/ngpus.

What is the num_batches?

----

`net.batch * net.subdivisions` = `batch_from_cfg`

https://github.com/AlexeyAB/darknet/blob/24c889d85705643ca4860dea88827a7f1893b731/src/parser.c#L618

----

The current iteration number is equal = `seen / batch_from_cfg`
https://github.com/AlexeyAB/darknet/blob/24c889d85705643ca4860dea88827a7f1893b731/src/network.c#L53

----

For 1 GPU will be loaded `batch_from_cfg` images and the `seen` will be increased by `batch_from_cfg`.
https://github.com/AlexeyAB/darknet/blob/24c889d85705643ca4860dea88827a7f1893b731/src/network_kernels.cu#L149
n-times, n = subdivisions: https://github.com/AlexeyAB/darknet/blob/24c889d85705643ca4860dea88827a7f1893b731/src/network.c#L322

----

For 4 GPU:
* will be loaded `batch_from_cfg * 4` images: https://github.com/AlexeyAB/darknet/blob/24c889d85705643ca4860dea88827a7f1893b731/src/detector.c#L72

* the `seen` will be increased by :
    ` *nets[0].seen += interval * n * nets[0].batch * nets[0].subdivisions = 4 * 4 * batch_from_cfg `

https://github.com/AlexeyAB/darknet/blob/24c889d85705643ca4860dea88827a7f1893b731/src/network_kernels.cu#L149
n-times, n = subdivisions: https://github.com/AlexeyAB/darknet/blob/24c889d85705643ca4860dea88827a7f1893b731/src/network.c#L322
+
https://github.com/AlexeyAB/darknet/blob/24c889d85705643ca4860dea88827a7f1893b731/src/network_kernels.cu#L363

**Comment by marvis** [comment]
Sorry, it is max_batches not num_batches.
https://github.com/pjreddie/darknet/blob/master/cfg/yolov3.cfg#L20

What I confused is that the total number of images processed between 1gpu and 4gpus is different.

For 1 gpu, the total number of images processed is num_batches * batch_from_cfg.
For 4 gpus, the total number of images processed is 4 * num_batches * batch_from_cfg.

There are four times difference. Am I understand right? 
And I think during 4gpus, the max_batches should be set to max_batches/4.

**Comment by AlexeyAB** [comment]
These are 2 ways:
* or decrease `max_batches `
* or increase iteration step (for example, increase iteration by 4 or by 16 for each batch)

----

For 1 GPU will be loaded `batch_from_cfg` images and the `seen` will be increased by `batch_from_cfg`.
For 4 GPU will be loaded `batch_from_cfg * 4` images and the `seen` will be increased by `4 * 4 * batch_from_cfg`

----

i.e.

For 1 GPU will be loaded `batch_from_cfg` images and the `iteration` will be increased by `1`.
For 4 GPU will be loaded `batch_from_cfg * 4` images and the `iteration` will be increased by `16`

So for 4 GPU: each GPU will process 16 times less images, and all 4 GPUs together will process 4 times less images than 1 GPU. I think therefore `learning_rate` is multipled by 4
